File: app/src/main/java/com/dss886/transmis/sms/SmsActivity.java
package com.dss886.transmis.sms;

import android.content.SharedPreferences;
import android.text.TextUtils;
import com.dss886.transmis.R;
import com.dss886.transmis.base.App;
import com.dss886.transmis.base.BaseActivity;
import com.dss886.transmis.utils.DialogBuilder;
import com.dss886.transmis.utils.Tags;
import com.dss886.transmis.view.SwitchItem;
import com.dss886.transmis.view.TextItem;

public class SmsActivity extends BaseActivity {

    private SwitchItem mSmsSwitch;
    private SwitchItem mMergeSwitch;
    private TextItem mTitleItem;
    private TextItem mContentItem;

    @Override
    protected int getToolbarTitle() {
        return R.string.sms_title;
    }

    @Override
    protected boolean showToolbarBackIcon() {
        return true;
    }

    @Override
    protected void addViews() {
        mSmsSwitch = new SwitchItem(this, "短信提醒", Tags.SP_SMS_ENABLE, true);
        mMergeSwitch = new SwitchItem(this, "合并长短信", Tags.SP_SMS_MERGE_LONG_TEXT, true);
        mTitleItem = new TextItem(this, "短信提醒标题").setCallback(sp -> {
            String value = sp.getString(Tags.SP_SMS_TITLE_REGEX, "默认");
            return TextUtils.isEmpty(value) ? "未设置" : value;
        });
        mContentItem = new TextItem(this, "短信提醒内容").setCallback(sp -> {
            String value = sp.getString(Tags.SP_SMS_CONTENT_REGEX, "默认");
            return TextUtils.isEmpty(value) ? "未设置" : value;
        });

        addView(mSmsSwitch);
        addView(mMergeSwitch);
        addView(mTitleItem);
        addView(mContentItem);
    }

    @Override
    protected void setListeners() {
        mSmsSwitch.setOnClickListener(v -> {});
        mMergeSwitch.setOnClickListener(v -> {});
        setTextItemListener(mTitleItem, Tags.SP_SMS_TITLE_REGEX, "设置提醒标题");
        setTextItemListener(mContentItem, Tags.SP_SMS_CONTENT_REGEX, "设置提醒内容");
    }

    private void setTextItemListener(TextItem item, String key, String showTitle) {
        item.setOnClickListener(v -> {
            String value = App.sp.getString(key, null);
            DialogBuilder.showEditTextDialog(this, showTitle, value, false, content -> {
                SharedPreferences.Editor editor = App.sp.edit();
                if (TextUtils.isEmpty(content)) {
                    editor.remove(key);
                } else {
                    editor.putString(key, content);
                }
                editor.apply();
                item.onResume();
            });
        });
    }
}

File: app/src/main/java/com/dss886/transmis/listen/SmsActivity.kt
package com.dss886.transmis.listen

import com.dss886.transmis.R
import com.dss886.transmis.base.App
import com.dss886.transmis.base.BaseSwitchActivity
import com.dss886.transmis.filter.FilterActivity
import com.dss886.transmis.filter.FilterType
import com.dss886.transmis.utils.Constants
import com.dss886.transmis.utils.DialogBuilder
import com.dss886.transmis.utils.TransmisManager
import com.dss886.transmis.utils.countOccurrences
import com.dss886.transmis.view.*
import java.util.*

class SmsActivity : BaseSwitchActivity() {

    private lateinit var mTitleConfig: EditTextConfig
    private lateinit var mContentConfig: EditTextConfig

    override fun getToolbarTitle(): String {
        return getString(R.string.sms_title)
    }

    override fun showToolbarBackIcon(): Boolean {
        return true
    }

    override fun getConfigs(): List<IConfig> {
        return mutableListOf<IConfig>().apply {
            add(SwitchConfig("短信开关", Constants.SP_SMS_ENABLE))
            add(SectionConfig("过滤"))
            add(TextConfig("发件人过滤", showRightArrow = true).apply {
                clickAction = {
                    FilterActivity.start(this@SmsActivity, FilterType.SMS_SENDER)
                }
                resumeAction = getResumeAction(this, FilterType.SMS_SENDER)
            })
            add(TextConfig("关键词过滤", showRightArrow = true).apply {
                clickAction = {
                    FilterActivity.start(this@SmsActivity, FilterType.SMS_KEYWORD)
                }
                resumeAction = getResumeAction(this, FilterType.SMS_KEYWORD)
            })
            add(SectionConfig("提醒设置"))
            add(EditTextConfig("提醒标题", Constants.SP_SMS_TITLE_REGEX).apply {
                isRequired = false
                mTitleConfig = this
            })
            add(EditTextConfig("提醒内容", Constants.SP_SMS_CONTENT_REGEX).apply {
                isRequired = false
                isMultiLine = true
                mContentConfig = this
            })
            add(InfoConfig(getString(R.string.info_sms_content)))
            add(SectionConfig("可选项"))
            add(SwitchConfig("合并长短信", Constants.SP_SMS_MERGE_LONG_TEXT))
            add(InfoConfig(getString(R.string.info_sms_merge_long)))
            add(SectionConfig("测试"))
            add(TestConfig("点击进行测试").apply {
                onTest = { doTest(this) }
            })
        }
    }

    private fun getResumeAction(config: TextConfig, type: FilterType): () -> Unit {
        val count = TransmisManager.getFilterCount(type)
        val text = if (count == 0) "无" else "$count 项"
        return {
            config.content = text
        }
    }

    private fun doTest(config: TestConfig) {
        config.content = "测试中"
        val title = mTitleConfig.getSpValue(null) ?: App.inst().getString(R.string.sms_title_default)
        var contentTemplate = mContentConfig.getSpValue(null)
        if (contentTemplate != null && contentTemplate.countOccurrences("%s") != 2) {
            config.content = "测试失败"
            config.reason = IllegalArgumentException("提醒内容参数有误")
            config.onFailure?.invoke()
            return
        }
        contentTemplate = contentTemplate ?: App.inst().getString(R.string.sms_content_default)
        val content = String.format(Locale.CHINA, contentTemplate, "123456789", "smsContent")
        DialogBuilder.showAlertDialog(this, title, content.trim(), null)
        config.content = "测试成功"
        config.onSuccess?.invoke()
    }

}